internal/runner/common/unit_filter_graph.go

```diff
@@ -0,0 +1,70 @@
+package common
+
+import (
+	"context"
+
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+)
+
+// UnitFilterGraph filters units to include only a target directory and its dependents.
+// This filter is used by the graph command to show only relevant units in the dependency graph.
+type UnitFilterGraph struct {
+	// TargetDir is the directory whose dependents should be included
+	TargetDir string
+}
+
+// Filter implements UnitFilter.
+// It marks all units as excluded except for the target directory and units that depend on it.
+func (f *UnitFilterGraph) Filter(ctx context.Context, units Units, opts *options.TerragruntOptions) error {
+	// Build dependency map first
+	dependentUnits := make(map[string][]string)
+
+	for _, unit := range units {
+		if len(unit.Dependencies) != 0 {
+			for _, dep := range unit.Dependencies {
+				dependentUnits[dep.Path] = util.RemoveDuplicatesFromList(append(dependentUnits[dep.Path], unit.Path))
+			}
+		}
+	}
+
+	// Recursively collect all dependent units
+	// Use a safe upper bound to prevent infinite loops
+	maxIterations := len(units)*len(units) + 1
```

Comment:
NIT: I don't understand why this is a safe upper bound (probably because I'm not thinking about this correctly). Unless this comment makes you reconsider that it is, please add a comment in the future as to why this is the safe upper bound here.